### PR TITLE
Support rdma write with immediate and write with completion WRs

### PR DESCRIPTION
**Pull request by rajrangoju**
This series adds support for rdma write with immediate and write with completion
 WRs.

These are generated on top of the following iw_cxgb4 submission
#359 #360 

Corresponding kernel submission for this PR is at
https://marc.info/?l=linux-rdma&m=153181311509759&w=2

### CODE REVIEW

**Comment by jgunthorpe** [comment]
Does this have anything to do with the srq stuff? Why is it based on the SRQ patches?

**Comment by bharatpotnuri** [comment]
> Does this have anything to do with the srq stuff? Why is it based on the SRQ patches?

@jgunthorpe These depend on 64Byte CQE support changes #359 . SRQ patches #360 are submitted before these and all these are underreview, so this PR referenced all of them.

Here is kernel patch submission for write with immediate and write with completion:
https://marc.info/?l=linux-rdma&m=153181311509759&w=2
https://marc.info/?l=linux-rdma&m=153181312009760&w=2
https://marc.info/?l=linux-rdma&m=153181312909765&w=2

**Comment by bharatpotnuri** [comment]
-> resolved merge conflicts.
-> updated immediate data structures to reflect the kernel mode changes for v3.

Posted v3 kernel patches:
https://patchwork.kernel.org/cover/10546467/
https://patchwork.kernel.org/patch/10546469/
https://patchwork.kernel.org/patch/10546471/

**Comment by bharatpotnuri** [comment]
-> resolved merge conflicts.
-> updated immediate data structures to reflect the kernel mode changes for v3.

Posted v3 kernel patches:
https://patchwork.kernel.org/cover/10546467/
https://patchwork.kernel.org/patch/10546469/
https://patchwork.kernel.org/patch/10546471/

File: providers/cxgb4/qp.c [hunk 1]
```diff
@@ -254,6 +270,58 @@ static int build_rdma_write(struct t4_sq *sq, union t4_wr *wqe,
...
+static void build_immd_cmpl(struct t4_sq *sq, struct fw_ri_immd_cmpl *immdp,
+			    struct ibv_send_wr *wr)
+{
+	memcpy((u8 *)immdp->data, (u8 *)(uintptr_t)wr->sg_list->addr, 16);
+	memset(immdp->r1, 0, 6);
+	immdp->op = FW_RI_DATA_IMMD;
+	immdp->immdlen = 16;
+}
+
+#define BUILD_BUG_ON(condition) ((void)sizeof(char[1 - 2 * !!(condition)]))
```
Comment (jgunthorpe) [inline providers/cxgb4/qp.c]: Nope, use C11 static_assert for this stuff.

Comment (bharatpotnuri) [reply providers/cxgb4/qp.c]: Replaced BUILD_BUG_ON() with static_assert()

Comment (bharatpotnuri) [reply providers/cxgb4/qp.c]: Generated kernel header cpmmit on top of `94245f4ad9e10c161affaa4763136d29fbe91cb9`.

**Pull request merged**